debian/bongooplog.1
.\" Man page generated from reStructuredText.
.
.TH "BONGOOPLOG" "1" "January 30, 2015" "3.0" "bongodb-manual"
.SH NAME
bongooplog \- BongoDB Oplog Utility
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.sp
New in version 2.2.

.SH SYNOPSIS
.sp
\fBbongooplog\fP is a simple tool that polls operations from
the \fIreplication\fP \fIoplog\fP of a remote server, and applies
them to the local server. This capability supports certain classes of
real\-time migrations that require that the source server remain online
and in operation throughout the migration process.
.sp
Typically this command will take the following form:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
bongooplog  \-\-from bongodb0.example.net \-\-host bongodb1.example.net
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This command copies oplog entries from the \fBbongod\fP instance
running on the host \fBbongodb0.example.net\fP and duplicates
operations to the host \fBbongodb1.example.net\fP\&. If you do not need
to keep the \fI\-\-from\fP host running during
the migration, consider using \fBbongodump\fP and
\fBbongorestore\fP or another \fBbackup\fP operation, which may be better suited to
your operation.
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
If the \fBbongod\fP instance specified by the \fI\-\-from\fP
argument is running with \fBauthentication\fP, then
\fBbongooplog\fP will not be able to copy oplog entries.
.UNINDENT
.UNINDENT
.sp
\fBSEE ALSO:\fP
.INDENT 0.0
.INDENT 3.5
\fBbongodump\fP, \fBbongorestore\fP,
http://docs.bongodb.org/manual/core/backups, http://docs.bongodb.org/manual/core/replica\-set\-oplog\&.
.UNINDENT
.UNINDENT
.SH OPTIONS
.sp
Changed in version 3.0.0: \fBbongooplog\fP removed the \fB\-\-dbpath\fP as well as related
\fB\-\-directoryperdb\fP and \fB\-\-journal\fP options. You must use
\fBbongooplog\fP while connected to a \fBbongod\fP instance.

.INDENT 0.0
.TP
.B bongooplog
.UNINDENT
.INDENT 0.0
.TP
.B bongooplog
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help
Returns information on the options and use of \fBbongooplog\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-verbose, \-v
Increases the amount of internal reporting returned on standard output
or in log files. Increase the verbosity with the \fB\-v\fP form by
including the option multiple times, (e.g. \fB\-vvvvv\fP\&.)
.UNINDENT
.INDENT 0.0
.TP
.B \-\-quiet
Runs the \fBbongooplog\fP in a quiet mode that attempts to limit the amount
of output.
.sp
This option suppresses:
.INDENT 7.0
.IP \(bu 2
connection accepted events
.IP \(bu 2
connection closed events
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-version
Returns the \fBbongooplog\fP release number.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-host <hostname><:port>, \-h <hostname><:port>
Specifies a resolvable hostname for the \fBbongod\fP instance to
which \fBbongooplog\fP will apply \fIoplog\fP operations
retrieved from the server specified by the \fI\-\-from\fP option.
.sp
By default \fBbongooplog\fP attempts to connect to a BongoDB instance running
on the localhost on port number \fB27017\fP\&.
.sp
To connect to a replica set, specify the \fBreplica set name\fP and a seed list of set members. Use the
following form:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
<replSetName>/<hostname1><:port>,<hostname2><:port>,<...>
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can always connect directly to a single BongoDB instance by
specifying the host and port number directly.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-port
Specifies the port number of the \fBbongod\fP instance where
\fBbongooplog\fP will apply \fIoplog\fP entries. Specify
this option only if the BongoDB instance to connect to is not
running on the standard port of \fB27017\fP\&. You may also specify a
port number using the \fI\-\-host\fP command.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ipv6
Enables IPv6 support and allows the \fBbongooplog\fP to connect to the
BongoDB instance using an IPv6 network. All BongoDB programs and
processes disable IPv6 support by default.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ssl
New in version 2.6.

.sp
Enables connection to a \fBbongod\fP or \fBbongos\fP that has
SSL support enabled.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCAFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains the root certificate chain
from the Certificate Authority. Specify the file name of the
\fB\&.pem\fP file using relative or absolute paths.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.sp
\fBWARNING:\fP
.INDENT 7.0
.INDENT 3.5
If the \fBbongo\fP shell or any other tool that connects to
\fBbongos\fP or \fBbongod\fP is run without
\fI\-\-sslCAFile\fP, it will not attempt to validate
server certificates. This results in vulnerability to expired
\fBbongod\fP and \fBbongos\fP certificates as well as to foreign
processes posing as valid \fBbongod\fP or \fBbongos\fP
instances. Ensure that you \fIalways\fP specify the CA file against which
server certificates should be validated in cases where intrusion is a
possibility.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains both the SSL certificate
and key. Specify the file name of the \fB\&.pem\fP file using relative
or absolute paths.
.sp
This option is required when using the \fI\-\-ssl\fP option to connect
to a \fBbongod\fP or \fBbongos\fP that has
\fBCAFile\fP enabled \fIwithout\fP
\fBallowConnectionsWithoutCertificates\fP\&.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyPassword <value>
New in version 2.6.

.sp
Specifies the password to de\-crypt the certificate\-key file (i.e.
\fI\-\-sslPEMKeyFile\fP). Use the \fI\-\-sslPEMKeyPassword\fP option only if the
certificate\-key file is encrypted. In all cases, the \fBbongooplog\fP will
redact the password from all logging and reporting output.
.sp
If the private key in the PEM file is encrypted and you do not specify
the \fI\-\-sslPEMKeyPassword\fP option, the \fBbongooplog\fP will prompt for a passphrase. See
\fIssl\-certificate\-password\fP\&.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCRLFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains the Certificate Revocation
List. Specify the file name of the \fB\&.pem\fP file using relative or
absolute paths.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidCertificates
New in version 2.6.

.sp
Bypasses the validation checks for server certificates and allows
the use of invalid certificates. When using the
\fBallowInvalidCertificates\fP setting, BongoDB logs as a
warning the use of the invalid certificate.
.sp
The default distribution of BongoDB does not contain support for SSL.
For more information on BongoDB and SSL, see http://docs.bongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidHostnames
New in version 3.0.

.sp
Disables the validation of the hostnames in SSL certificates. Allows
\fBbongooplog\fP to connect to BongoDB instances if the hostname their
certificates do not match the specified hostname.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslFIPSMode
New in version 2.6.

.sp
Directs the \fBbongooplog\fP to use the FIPS mode of the installed OpenSSL
library. Your system must have a FIPS compliant OpenSSL library to use
the \fI\-\-sslFIPSMode\fP option.
.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
FIPS Compatible SSL is
available only in \fI\%BongoDB Enterprise\fP\&. See
http://docs.bongodb.org/manual/tutorial/configure\-fips for more information.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-username <username>, \-u <username>
Specifies a username with which to authenticate to a BongoDB database
that uses authentication. Use in conjunction with the \fB\-\-password\fP and
\fB\-\-authenticationDatabase\fP options.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-password <password>, \-p <password>
Specifies a password with which to authenticate to a BongoDB database
that uses authentication. Use in conjunction with the \fB\-\-username\fP and
\fB\-\-authenticationDatabase\fP options.
.sp
If you do not specify an argument for \fI\-\-password\fP, \fBbongooplog\fP will
prompt interactively for a password on the console.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-authenticationDatabase <dbname>
New in version 2.4.

.sp
Specifies the database that holds the user\(aqs credentials.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-authenticationMechanism <name>
\fIDefault\fP: BONGODB\-CR
.sp
New in version 2.4.

.sp
Changed in version 2.6: Added support for the \fBPLAIN\fP and \fBBONGODB\-X509\fP authentication
mechanisms.

.sp
Specifies the authentication mechanism the \fBbongooplog\fP instance uses to
authenticate to the \fBbongod\fP or \fBbongos\fP\&.
.TS
center;
|l|l|.
_
T{
Value
T}	T{
Description
T}
_
T{
BONGODB\-CR
T}	T{
BongoDB challenge/response authentication.
T}
_
T{
BONGODB\-X509
T}	T{
BongoDB SSL certificate authentication.
T}
_
T{
PLAIN
T}	T{
External authentication using LDAP. You can also use \fBPLAIN\fP
for authenticating in\-database users. \fBPLAIN\fP transmits
passwords in plain text. This mechanism is available only in
\fI\%BongoDB Enterprise\fP\&.
T}
_
T{
GSSAPI
T}	T{
External authentication using Kerberos. This mechanism is
available only in \fI\%BongoDB Enterprise\fP\&.
T}
_
.TE
.UNINDENT
.INDENT 0.0
.TP
.B \-\-gssapiServiceName
New in version 2.6.

.sp
Specify the name of the service using \fBGSSAPI/Kerberos\fP\&. Only required if the service does not use the
default name of \fBbongodb\fP\&.
.sp
This option is available only in BongoDB Enterprise.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-gssapiHostName
New in version 2.6.

.sp
Specify the hostname of a service using \fBGSSAPI/Kerberos\fP\&. \fIOnly\fP required if the hostname of a machine does
not match the hostname resolved by DNS.
.sp
This option is available only in BongoDB Enterprise.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-db <database>, \-d <database>
Specifies the name of the database on which to run the \fBbongooplog\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-collection <collection>, \-c <collection>
Specifies the collection to export.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-seconds <number>, \-s <number>
Specify a number of seconds of operations for \fBbongooplog\fP to
pull from the \fIremote host\fP\&. Unless
specified the default value is \fB86400\fP seconds, or 24 hours.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-from <host[:port]>
Specify the host for \fBbongooplog\fP to retrieve \fIoplog\fP
operations from. \fBbongooplog\fP \fIrequires\fP this option.
.sp
Unless you specify the \fI\-\-host\fP option,
\fBbongooplog\fP will apply the operations collected with this
option to the oplog of the \fBbongod\fP instance running on the
localhost interface connected to port \fB27017\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-oplogns <namespace>
Specify a namespace in the \fI\-\-from\fP host
where the oplog resides. The default value is \fBlocal.oplog.rs\fP, which
is the where \fIreplica set\fP members store their operation log.
However, if you\(aqve copied \fIoplog\fP entries into another
database or collection or are pulling oplog entries from a
\fBmaster\-slave\fP deployment, use \fI\%\-\-oplogns\fP to
apply oplog entries stored in another location. Namespaces take the
form of \fB[database].[collection]\fP\&.
.UNINDENT
.SH USE
.sp
Consider the following prototype \fBbongooplog\fP command:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
bongooplog  \-\-from bongodb0.example.net \-\-host bongodb1.example.net
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Here, entries from the \fIoplog\fP of the \fBbongod\fP running
on port \fB27017\fP\&. This only pull entries from the last 24 hours.
.sp
Use the \fI\-\-seconds\fP argument to capture
a greater or smaller amount of time. Consider the following example:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
bongooplog  \-\-from bongodb0.example.net \-\-seconds 172800
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
In this operation, \fBbongooplog\fP captures 2 full days of
operations. To migrate 12 hours of \fIoplog\fP entries, use the
following form:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
bongooplog  \-\-from bongodb0.example.net \-\-seconds 43200
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
BongoDB Documentation Project
.SH COPYRIGHT
2011-2015
.\" Generated by docutils manpage writer.
.


NEWS
WiredTiger release 2.9.0, 2016-09-06
------------------------------------

New features and API changes; refer to the API documentation for full details:

* WT-2360	Add nesting and disjunction functionality to cursor joins
* WT-2552	Add a public API to allow custom filesystem implementations
* WT-2738	Remove the ability to change the default checkpoint name. See the upgrading documentation for details.
* WT-2711	Change the statistics log configuration options. Restrict where statistics log files can be created to remove security vulnerability. See upgrading documentation for details.
* WT-2764	Enhance the checkpoint implementation, which led to a change in default eviction settings. See the upgrading documentation for details.
* WT-2880	Add Zstandard compression support
* See the upgrading documentation for information about renamed and removed statistics.

Significant changes and bug fixes:

* WT-2026	Fix a bug where in-memory pages were allowed to grow too large
* WT-2343	Ensure we cannot rename/drop when a backup cursor is open
* WT-2520	Ensure WT_SESSION::verify does not alter tables
* WT-2576	Ensure variable-length column-store cannot out-of-order return
* WT-2664	Change eviction so any eviction thread can find candidates
* WT-2702	Resolve an issue where under high thread load, WiredTiger exceeds cache size
* WT-2731	Resolve an issue where raw compression can create pages that are larger than expected
* WT-2737	Allow reconciliation to scrub dirty pages rather than evicting them
* WT-2757	Fix a bug in column stores where it was not possible to retrieve the new record number after an append
* WT-2816	Improve eviction performance

Issues fixed in MongoDB:

* SERVER-23504	Resolve a resource leak
* SERVER-23588	Stop using _open_osfhandle on Windows
* SERVER-23659	Improve log error message at startup
* SERVER-23661	Fix an issue where $sample takes disproportionately long time on newly created collection
* SERVER-24306	Fix stall in log_flush switching to new files
* SERVER-24580	Add more eviction stats to track efficiency
* SERVER-24580	Enhance eviction when application threads are contributing
* SERVER-24971	Don't cache buffers after application eviction
* SERVER-25843	Remove a redundant test
* SERVER-25845	Resolve an explicit null dereferenced
* SERVER-25846	Resolve a dereference after null with a check
* SERVER-26753	Don't spin on a read-lock in a tight loop

Other noteworthy changes since the previous release:

* WT-2103	Add incremental backup testing to test/format
* WT-2330	Ensure In-memory configurations do not create on-disk collection files
* WT-2450	Resolve an issue where salvage releases pages, then explicitly evicts them. Which can race with the eviction server
* WT-2453	Add multiple eviction queues
* WT-2504	Remove READONLY conditional for base config
* WT-2505	Resolve clang analyzer warnings
* WT-2508	Ensure test programs should remove test directories on the "clean" target
* WT-2512	Change how wtperf throttle values per thread are calculated
* WT-2513	Fix a type conversion warning
* WT-2517	Resolve incorrect uses of setvbuf on Windows
* WT-2518	Optimization of LSM checkpoint handle acquisition
* WT-2522	Resolve incorrect format code in message
* WT-2525	Miscellaneous cleanups of In-memory configurations
* WT-2526	Prevent mixing and matching readonly file-handles and read/write data-handles
* WT-2528	Fix a style error in WiredTiger build
* WT-2529	Be less aggressive asserting in readonly connections
* WT-2531	Resolve a case where in-memory tables are wasting space in truncation
* WT-2532	Change the WT_STREAM_APPEND and WT_STREAM_LINE_BUFFER flags as they overlap
* WT-2533	Don't let in-memory tables return a zero size
* WT-2534	Use atomic add when allocating transaction IDs
* WT-2535	Extend test/format to test for transactions reading their writes
* WT-2537	Check for an old and new uninitialized LSN for recovery
* WT-2539	Make streams a separate handle type from files
* WT-2540	Separate stream and file handle methods
* WT-2542	Resolve fixed-length column store reconciliation overwriting original values
* WT-2544	Fix eviction stats when clear is used
* WT-2546	Change evict server to be smarter when it decides whether or not to evict pages
* WT-2547	Add eviction testing configurations for 1 thread
* WT-2548	Cap the amount of data handed to raw compression
* WT-2550	Add support for raw mode cursors in java
* WT-2553	Add some In-memory documentation and stress testing
* WT-2556	Fix a typo in the Java example code
* WT-2557	Ensure that we truncate logs before closing backup cursor
* WT-2558	Reorganize WT_PAGE structure
* WT-2559	Resolve a potential segfault in Windows logging code
* WT-2560	Add a rwlock to protect transaction state, don't spin on this lock
* WT-2562	Add sleep loop to adapt for slow machines like PPC
* WT-2565	Further limit the amount of data passed to raw compression
* WT-2566	Add explicit memory barriers to Lock/unlock operations
* WT-2567	Add a check to see if logging is enabled for truncate rather than assuming it is
* WT-2568	Add a backward compatible constructor for PackInputStream
* WT-2569	Ensure __win_handle_read is followed by a call to GetLastError()
* WT-2570	Minor lint and cleanups to code
* WT-2571	Join code clean up
* WT-2572	Ensure test/format doesn't select an incompatible set of options for in-memory options configured
* WT-2573	Resolve a potential case where a double free can occur
* WT-2574	Ensure test/format frees all allocated configure memory
* WT-2576	Ensure variable-length column-store cannot out-of-order return
* WT-2577	Resolve a potential double-free following an in-memory split
* WT-2579	Change so diagnostic only code can use file streams directly
* WT-2580	Fix potential SWIG naming conflict in Java
* WT-2581	Resolve a case where we fail to track saved updates during a page split
* WT-2582	Fix a case where eviction can return WT_RESTART when running in-memory
* WT-2584	Change to not use periods at the end of error messages

src/bongo/gotools/bongooplog/main/bongooplog.go
// Main package for the bongooplog tool.
package main

import (
	"github.com/bongodb/bongo-tools/common/db"
	"github.com/bongodb/bongo-tools/common/log"
	"github.com/bongodb/bongo-tools/common/options"
	"github.com/bongodb/bongo-tools/common/signals"
	"github.com/bongodb/bongo-tools/common/util"
	"github.com/bongodb/bongo-tools/bongooplog"
	"os"
)

func main() {
	// initialize command line options
	opts := options.New("bongooplog", bongooplog.Usage,
		options.EnabledOptions{Auth: true, Connection: true, Namespace: false})

	// add the bongooplog-specific options
	sourceOpts := &bongooplog.SourceOptions{}
	opts.AddOptions(sourceOpts)

	log.Logvf(log.Always, "warning: bongooplog is deprecated, and will be removed completely in a future release")

	// parse the command line options
	args, err := opts.Parse()
	if err != nil {
		log.Logvf(log.Always, "error parsing command line options: %v", err)
		log.Logvf(log.Always, "try 'bongooplog --help' for more information")
		os.Exit(util.ExitBadOptions)
	}

	if len(args) != 0 {
		log.Logvf(log.Always, "positional arguments not allowed: %v", args)
		log.Logvf(log.Always, "try 'bongooplog --help' for more information")
		os.Exit(util.ExitBadOptions)
	}

	// print help, if specified
	if opts.PrintHelp(false) {
		return
	}

	// print version, if specified
	if opts.PrintVersion() {
		return
	}

	// init logger
	log.SetVerbosity(opts.Verbosity)
	signals.Handle()

	// connect directly, unless a replica set name is explicitly specified
	_, setName := util.ParseConnectionString(opts.Host)
	opts.Direct = (setName == "")
	opts.ReplicaSetName = setName

	// validate the bongooplog options
	if sourceOpts.From == "" {
		log.Logvf(log.Always, "command line error: need to specify --from")
		os.Exit(util.ExitBadOptions)
	}

	// create a session provider for the destination server
	sessionProviderTo, err := db.NewSessionProvider(*opts)
	if err != nil {
		log.Logvf(log.Always, "error connecting to destination host: %v", err)
		os.Exit(util.ExitError)
	}
	defer sessionProviderTo.Close()

	// create a session provider for the source server
	opts.Connection.Host = sourceOpts.From
	opts.Connection.Port = ""
	sessionProviderFrom, err := db.NewSessionProvider(*opts)
	if err != nil {
		log.Logvf(log.Always, "error connecting to source host: %v", err)
		os.Exit(util.ExitError)
	}
	defer sessionProviderFrom.Close()

	// initialize bongooplog
	oplog := bongooplog.BongoOplog{
		ToolOptions:         opts,
		SourceOptions:       sourceOpts,
		SessionProviderFrom: sessionProviderFrom,
		SessionProviderTo:   sessionProviderTo,
	}

	// kick it off
	if err := oplog.Run(); err != nil {
		log.Logvf(log.Always, "error: %v", err)
		os.Exit(util.ExitError)
	}

}
